docs/src/main/asciidoc/se/injection.adoc

```diff
@@ -0,0 +1,677 @@
+
+- Create a new ServiceRegistry instance and search for the desired service.
+- Inject the existing ServiceRegistry instance currently used by Helidon and retrieve services from it.
+
+To create a registry instance:
+
+[source,java]
+----
+include::{sourcedir}/se/inject/ServiceRegistryExample.java[tag=snippet_1, indent=0]
+----
+
+Note that all instances are created lazily, meaning the service registry does nothing by default.
+If a service performs any actions during construction or post-construction,
+you must first retrieve an instance from the registry to trigger its initialization.
+
+Special registry operations:
+
+- `List<ServiceInfo> lookupServices(Lookup lookup)` - get all service descriptors that match the lookup
+- `Optional<T> get(ServiceInfo)` - get an instance for the provided service descriptor
+
+The common registry operations are grouped by method name. Acceptable parameters are described below.
+
+Registry methods:
+
+- `T get(...)` - immediately get an instance of a contract from the registry; throws if implementation not available
+- `Optional<T> first(...)` - immediately get an instance of a contract from the registry; there may not be an implementation
+available
+- `List<T> all(...)` - immediately get all instances of a contract from the registry; result may be empty
+- `Supplier<T> supply(...)` - get a supplier of an instance; the service may be instantiated only when `get` is called
+- `Supplier<Optional<T>> supplyFirst(...)` - get a supplier of an optional instance
+- `Supplier<List<T>> supplyAll(...)` - get a supplier of all instances
+
+Lookup parameter options:
+
+- `Class<?>` - the contract we are looking for
+- `TypeName` - the same, but using Helidon abstraction of type names (may have type arguments)
+- `Lookup` - a full search criteria for a registry lookup
+
+== Startup
+
+Helidon provides a Maven plugin (`io.helidon.service:helidon-service-maven-plugin`, goal `create-application`) to generate
+build time bindings, that can be used to start the service registry without any classpath discovery and reflection.
+Default name is `ApplicationBinding` (customizable)
+
+Methods that accept the bindings are on `io.helidon.service.registry.ServiceRegistryManager`:
+
+- `start(Binding)` - starts the service registry with the generated binding, initializing all singleton and per-lookup services annotated with a `@RunLevel` annotation (i.e. `start(ApplicationBinding.create())`)
+- `start(Binding, ServiceRegistryConfig)` - same as above, allows for customization of configuration, if used, do not forget to set discovery to `false` to prevent automated discovery from the classpath
+
+All options to start a Helidon application that uses service registry:
+
+- A custom Main method using `ServiceRegistryManager.start(...)` methods, or `ServiceRegistryManager.create(...)` methods
+- A generated `ApplicationMain` - optional feature of the Maven plugin, requires property `generateMain` to be set to `true`
```

Comment:
also mention it uses `RunLevel` actively, but via code generated class.
This is the only approach that is fully reflection free and skips lookups for injection points.